src/Data/Text/Markdown.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Data.Text.Markdown where
+
+import CMarkGFM qualified as CMark
+import Control.Applicative ((<|>))
+import Data.Function ((&))
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Yaml
+import Swarm.Language.Parse (readTerm)
+import Swarm.Language.Pretty (prettyText)
+import Swarm.Language.Syntax (Syntax)
+import Swarm.Language.Pipeline (processParsedTerm, ProcessedTerm (..), prettyTypeErr)
+import Swarm.Language.Module (moduleAST)
+import Control.Monad (void)
+
+data Node
+  = Node NodeType [Node]
+  | LeafText Text
+  | LeafCode Syntax
+  | LeafCodeBlock String Syntax
+  deriving (Eq, Show)
+
+data NodeType
+  = Document
+  | Paragraph
+  deriving (Eq, Show)
+
+instance ToJSON Node where
+  toJSON = String . toText
+
+instance FromJSON Node where
+  parseJSON v =
+    withArray "markdown paragraphs" (foldMap parseJSON) v
+      <|> withText "markdown text" fromTextM v
+instance Semigroup Node where
+  Node Document cs1 <> Node Document cs2 = Node Document (cs1 <> cs2)
+  Node Document cs1 <> n2 = Node Document (cs1 <> [n2])
+  n1 <> Node Document cs2 = Node Document (n1 : cs2)
+  n1 <> n2 = Node Document [n1, n2]
+
+instance Monoid Node where
+  mempty = Node Document []
+
+fromText :: Text -> Node
+fromText = either error id . fromTextE
+
+fromTextM :: MonadFail m => Text -> m Node
+fromTextM = either fail pure . fromTextE
+
+fromTextE :: Text -> Either String Node
+fromTextE = fromGeneralNode . CMark.commonmarkToNode [CMark.optSourcePos] []
+
+-- TODO: use transformer kind of Error? (report all errors)
+fromGeneralNode :: CMark.Node -> Either String Node
+fromGeneralNode (CMark.Node p t cs) = case t of
+  CMark.DOCUMENT -> Node Document <$> mapM fromGeneralNode cs
+  CMark.PARAGRAPH -> Node Paragraph <$> mapM fromGeneralNode cs
+  CMark.SOFTBREAK -> leaf $ LeafText "\n" -- ?
+  CMark.LINEBREAK -> leaf $ LeafText "\n"
+  CMark.TEXT txt -> leaf $ LeafText txt
+  CMark.CODE txt -> leaf' $ LeafCode <$> parse txt
+  CMark.CODE_BLOCK i txt -> leaf' $ LeafCodeBlock (T.unpack i) <$> parse txt
+  _ -> Left $ "unsupported markdown node at " <> maybe "???" show p <> ":" <> show t
+ where
+  parse txt =
+    readTerm txt & \case
+      Left e -> Left (T.unpack e)
+      Right Nothing -> Left "empty code"
+      Right (Just s) -> case processParsedTerm s of
```

Comment:
We currently use inline code for types like `dir -> cmd unit`.

I think I can solve the type signatures as an alternative parse.

Reply:
Yeah, we can make a parallel to `readTerm` for types, like `readType = runParser (fully parseType)`.  If `readTerm` fails you can just try running `readType` too.